File: Python/utils/exercise_1_chirp.py
from textwrap import dedent
from .question import NoSolveQuestion


"""
import matplotlib.pyplot as plt  # for plotting
import numpy as np               # for numerics
import qutip                     # for quantum mechanics
from numpy import pi, sqrt, exp, sin, cos
"""

problem_1 = NoSolveQuestion(
    solution_message=dedent(
        r"""
        ```python
        t = np.linspace(-250, 250, 5000)
        ```
        """
    ),
    hint=dedent(
        r"""
        You can use NumPy's function `np.linspace()`, which needs the
        starting point, the end point and the number of gridpoints as
        arguments.
        """
    ),
)


problem_2 = NoSolveQuestion(
    solution_message=dedent(
        r"""
        ```python
        H1 = -1/2 * qutip.sigmaz()
        H2 = -1/2 * qutip.sigmax()
        ```
        """
    ),
    hint=dedent(
        r"""
        You can convert a NumPy array to a quantum object for example
        by writing `qutip.Qobj(np.array[1, 0])`.

        Alternatively you can write the Hamiltonian in terms of Pauli
        matrices, given by `qutip.sigmax()`, `qutip.sigmay()` and
        `qutip.sigmaz()`.
        """
    ),
)


problem_3 = NoSolveQuestion(
    solution_message=dedent(
        r"""
        ```python
        E0 = 1
        tau = 50
        alpha = 0.001

        Env = E0 * np.exp(-t**2/(2*tau**2))
        dϕ = alpha * t
        ```
        """
    ),
    hint=dedent(
        r"""
        You can multiply to variables with the `*` operator,
        whereas `t**2` corresponds to "t to the power of two".

        The import statements above defined the short hand notation
        `cos`, `exp` etc. for the NumPy functions `np.cos`, `np.exp`.
        """
    ),
)


"""
H = [[H1, dϕ], [H2, Env]]

def evaluate(H, time_index):
    (H1, dϕ), (H2, Env) = H
    return dϕ[time_index] * H1 + Env[time_index] * H2
"""

problem_4 = NoSolveQuestion(
    solution_message=dedent(
        r"""
        ```python
        psi0 = qutip.basis(2, 0)
        psi1 = qutip.basis(2, 1)
        ```
        """
    ),
    hint=dedent(
        r"""
        Again, you can use `qutip.Qobj` to convert NumPy arrays or
        lists to quantum objects.

        Alternatively, you can produce the states with
        `qutip.basis(dim, n)`, which created the `n`th basis state
        of dimension `dim`.
        """
    ),
)


problem_5 = NoSolveQuestion(
    solution_message=dedent(
        r"""
        ```python
        proj0 = psi0 * psi0.dag()
        proj1 = psi1 * psi1.dag()
        ```
        """
    ),
    hint=dedent(
        r"""
        You can turn a ket state `psi` defined as a `qutip.Qobj` into
        a bra by writing `psi.dat()`.

        Alternatively, write `psi.proj()` to directly create a
        projector from this state.
        """
    ),
)


problem_6 = NoSolveQuestion(
    solution_message=dedent(
        r"""
        ```python
        output = qutip.mesolve(H, psi0, t, e_ops=[proj0, proj1])
        ```
        """
    ),
    hint=dedent(
        r"""
        You need to specify the following arguments to `qutip.mesolve()`
        - Hamiltonian (as nested list)
        - initial state
        - time grid
        - observables: these are given in the form
        `e_ops=[op1, op2, ...]`.
        """
    ),
)

problem_7 = NoSolveQuestion(
    solution_message=dedent(
        r"""
        ```python
        fig, ax = plt.subplots()
        ax.plot(t, np.abs(Env)/np.abs(Env).max(), '-',
                color='lightgrey', label='|E|')
        ax.plot(t, output.expect[0], '-', label=r'$|0\rangle$')
        ax.plot(t, output.expect[1], '--', label=r'$|1\rangle$')
        ax.set_xlim(t.min(), t.max())
        ax.set_xlabel('Time')
        ax.set_ylabel('Population')
        ax.legend()
        plt.show()
        ```
        """
    ),
    hint=dedent(
        r"""
        The time-dependent population of the `i`th state is stored in
        `output.expect[i]`.

        You can create a quick and dirty plot of `f` over `x` with
        `plt.plot(x, f)`.
        Check the documentation of the Matplotlib package for in depth
        plotting tutorials.
        """
    ),
)


bonus = NoSolveQuestion(
    solution_message=dedent(
        r"""
        ```python
        H = (np.array(H1) * dϕ[:,np.newaxis,np.newaxis]
             + np.array(H2) * Env[:,np.newaxis,np.newaxis])

        e, v = np.linalg.eig(H)

        # Optional: fix ordering of eigenvalues for t>0
        e[t>0] = e[t>0][:, ::-1]
        v[t>0] = np.swapaxes(v[t>0][:, ::-1], -1, -2)

        fig, ax = plt.subplots()
        ax.plot(t, -dϕ/2, '-', color='grey')
        ax.plot(t, +dϕ/2, '--', color='grey')
        ax.plot(t, e[:, 0].real, '-', label='state 0')
        ax.plot(t, e[:, 1].real, '--', label='state 1')
        ax.set_xlabel('t')
        ax.set_ylabel('Eigenenergie')
        ax.legend()
        plt.show()
        ```
        """
    ),
    hint=dedent(
        r"""
        Construct the total Hamiltonian on the time grid as
        3d NumPy array with dimension `(Nt, 3, 3)`, where `Nt` is the
        number of points in the time grid.
        You can convert the previously defined quantum objects `H1` and `H2`
        to NumPy arrays with `np.array()` and multiply them with `Env` and `dϕ`
        by using NumPy's broadcasting rules and `np.newaxis`.

        Diagonalise the time-dependent Hamiltonian with NumPy's function
        `np.linalg.eig()`.
        """
    ),
)

File: Julia/utils/exercise_1_chirp.jl
include("question.jl")

#=
using QuantumPropagators: QuantumPropagators, hamiltonian, propagate
using OrdinaryDiffEq
using Plots

Plots.default(
    linewidth               = 2.0,
    foreground_color_legend = nothing,
    background_color_legend = RGBA(1, 1, 1, 0.8)
)
=#

problem_1 = Question(
    hint="""
      **Hint**: Combine the built-in functions `range` and `collect`.
      """,
    solution="""
      **Solution**:

      ```julia
      t = collect(range(-250, 250; length=5000));
      ```
      """,
)

problem_2 = Question(
    hint="""
      **Hint**:
      Matrices in Julia can be written by separating
      columns with spaces and rows with a semicolon or a newline (for a
      multiline format)
      """,
    solution=raw"""
      **Solution**:
      ```julia
      H₁ = -1/2 * [
          1  0
          0 -1
      ]
      H₂ = -1/2 * [
          0 1
          1 0
      ]
      ```

      or

      ```julia
      H₁ = -1/2 * [1 0; 0 -1]
      H₂ = -1/2 * [0 1; 1  0 ]
      ```
      """,
)

problem_3 = Question(
    hint=raw"""
      You can multiply to variables with the `*` operator,
      whereas `t^2` corresponds to "t to the power of two". The `*` can be
      omitted between numbers and variables, so `2τ` is equivalent to
      `2 * τ`.

      For unicode symbols, write latex-like code and convert it to unicode
      with the `<tab>` key.

      * `E\_0<tab>` becomes `E₀`
      * `\tau<tab>` becomes `τ`
      * `\alpha<tab>` becomes `\alpha`
      * `\phi<tab>` becomes `ϕ`

      You can always paste a particular unicode symbol into the help system
      to see how it can be written:

      ```
      ? ϕ
      ```
      """,
    solution="""
      **Solution**:

      ```julia
      E₀ = 1.0
      τ = 50.0
      α = 0.001

      Env(t) = E₀ * exp(-t^2 / 2τ^2)
      dϕ(t) = α * t
      ```
      """,
)

#=
H = hamiltonian((H₁, dϕ), (H₂, Env))

using QuantumPropagators.Controls: evaluate
=#

problem_4 = Question(
    hint=raw"""
        **Hint**:

        * An array with elements of type `Type` is create in Julia with
          `Type[…]`, e.g., `Float64[1.0, 1.1]`, where `Float64` is a
          double-precision (64 bit) float
        * When no type is given, Julia automatically derives the type from the
          values, e.g., `[1, 2]` is the same as `Int64[1, 2]`.
        * The type of a complex (double-precision) number is `ComplexF64`
        * Quantum state vectors have complex values
        """,
      solution="""
      **Solution**:

      ```
      Ψ₀ = ComplexF64[1, 0]
      Ψ₁ = ComplexF64[0, 1]
      ```
      """,
)

problem_5 = Question(
    hint=raw"""
        **Hint**:
        Julia uses `'` for complex conjugation of vectors and matrices, as a
        short-hand for `LinearAlgebra.Adjoint()`.
        """,
    solution="""
        **Solution**:

        ```julia
        P₀ = Ψ₀ * Ψ₀'
        P₁ = Ψ₁ * Ψ₁'
        ```
        """,
)

problem_6 = Question(
    hint=raw"""
        **Hint**

        You need to specify the following arguments to
        `QuantumPropagators.propagate`:

        As positional arguments:

        - initial state
        - Hamiltonian
        - time grid

        As keyword arguments

        - method: the propagation method (`OrdinaryDiffEq`)
        - observables: these are given in the form
        `observables=[op1, op2, ...]`.

        By default, `propagate` returns the final propagated states. Here, we
        want the expectation values of the projectores `P₀` and `P₁` for each
        point in time. We can obtain these by passing `storage=true`

        For details, query the help system about the `propagate` function
        (`? propagate`).
        """,
    solution=raw"""
        **Solution**:

        ```julia
        output = propagate( Ψ₀, H, t; method=OrdinaryDiffEq, observables=[P₀, P₁], storage=true )
        ```
        """,
)

problem_7 = Question(
    hint=raw"""
        **Hint**

        * When plotting the function `Env`, map it elementwise to the vector `t`
          using Julia's dot feature (`abs.(Env.(t))`). Also normalize to the peak
          value `maximum(abs.(Env.(t)))`
        * Use the `plot` function for the initial curve, giving an array of
          x-values and an array of y-values, and the options like color or
          label as keyword arguments (separated from the positional arguments
          by a semicolon)
        * Add more lines to the plot via the `plot!` function. In general, an
          exclamation mark at the end of a function in Julia is a convention
          to indicate that a function modifies an object. In this case, that
          object is the current plot (a hidden internal object)
        * You can use a call to `plot!` with only keyword arguments to set
          general properties like the axis labels. Don't forget the semicolon,
          or the `Plots` package might get confused about what parameters are
          data to be plotted and what parameters are options.
        """,
    solution=raw"""
        **Solution**:

        ```julia
        E_max = maximum(abs.(Env.(t)))
        plot(t, abs.(Env.(t)) / E_max; color="lightgray", label="|ℰ|")
        plot!(t, real.(output[1,:]); color="#1f77b4", label="|0⟩")
        plot!(t, real.(output[2,:]); color="#ff7f0e", linestyle=:dash, label="|1⟩")
        plot!(; xlabel="Time", ylabel="Population", legend=:right )
        ```
        """,
)

bonus = Question(
    hint=raw"""
        **Hint**:

        * Create a list of matrices, containing `H` evaluated for each point of
          the time grid
        * Map the function `eigvals` to that list of matrices to get a list of
          lists of eigenvalues at each point in time (two eigenvalues at `t`).
          Remember that any function can be mapped to an array by appending a
          dot (`func.(array)` is equivalent to `[func(val) for val in array]`)
        * Plot the the two eigenvalues over time. Also plot `±dϕ(t)` for
          comparison.
        """,
    solution="""
        **Solution**:

        ```julia
        H_over_t = [Array(evaluate(H, tval)) for tval in t]
        evals = eigvals.(H_over_t)

        plot(t, t -> -dϕ(t)/2; color="grey", linestyle=:dash, label="")
        plot!(t, t ->  dϕ(t)/2; color="grey", linestyle=:dash, label="")
        plot!(t, [e[1] for e in evals]; color="#1f77b4", label="E₀")
        plot!(t, [e[2] for e in evals]; color="#ff7f0e", label="E₁")
        plot!(; xlabel="t", ylabel="Eigenenergies" )
        ```
        """,
)


"""
# Super-Bonus: Track the eigenstates
#
# It's the eigenvalues that should vary continuously, and in theory, we need to
# make sure that the eigenvalues correspond to the "correct" eigenstate, i.e.,
# that the eigenstates didn't "flip" at the avoided crossing.

using LinearAlebra

eigensys = LinearAlgebra.eigen.(H_over_t)

# How much do the eigenvectors change compared to the previous point in time?
function evec_change(U1, U2)
    [1 - abs2(U1[:,i] ⋅ U2[:,i]) for i=1:2]
end

# Do the eigenvectors "flip"?
function evec_flipped(U1, U2)
    [1 - abs2(U1[:,1] ⋅ U2[:,2]),  1 - abs2(U1[:,2] ⋅ U2[:,1])]
end

change = hcat([evec_change(eigensys[i].vectors, eigensys[i-1].vectors) for i=2:length(eigensys)]...)

plot(t[2:end], change')

# If there's a discontinuity in that plot, we're in trouble (but there's not)

flipped = hcat([evec_flipped(eigensys[i].vectors, eigensys[i-1].vectors) for i=2:length(eigensys) ]...)

plot(t[2:end], flipped')
"""
